NDTensors/src/abstractarray/set_types.jl

```diff
@@ -37,30 +37,34 @@ function set_indstype(arraytype::Type{<:AbstractArray}, dims::Tuple)
   return set_ndims(arraytype, length(dims))
 end
 
-function set_eltype_if_unspecified(
+function specify_eltype(
   arraytype::Type{<:AbstractArray{T}}, eltype::Type=default_eltype()
 ) where {T}
   return arraytype
 end
 
 #TODO transition to set_eltype when working for wrapped types
-function set_eltype_if_unspecified(
-  arraytype::Type{<:AbstractArray}, eltype::Type=default_eltype()
+function specify_eltype(arraytype::Type{<:AbstractArray}, eltype::Type=default_eltype())
+  return similartype(arraytype, eltype)
+end
+
+function specify_eltype(
+  arraytype::Type{<:AbstractArray{UnspecifiedZero}}, eltype::Type=default_eltype()
```

Comment:
Ideally we wouldn't need overloads like this, and we could "register" `UnspecifiedZero` as an unspecified parameter like `UnspecifiedParameter` by overloading functionality of `SetParameters` directly.

I think we just need to overload this function: https://github.com/ITensor/ITensors.jl/blob/c49da9e02f7c2465001f3a8446ad49faf23347d0/NDTensors/src/SetParameters/src/set_unspecified_parameters.jl#L7-L11, i.e.
```julia
function SetParameters.replace_unspecified_parameters(
  type::Type, position::Position, current_parameter::Type{<:UnspecifiedZero}, parameter
)
  return set_parameters(type, position, parameter)
end
```